File: api_helpers/python/passivetotal.py
#!/usr/bin/env python

"""
Copyright (c) 2015, PassiveTotal LLC.
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
"""

import requests
import json
import logging
import sys

class set_check(object):
	"""Python decorator to perform field checks and input tests
	for calls where data is being POSTed to the API"""
	def __init__(self, field, test):
		self.field = field
		self.test = test
	def __call__(self, func):
		def wrapped(*args, **kwargs):
			if self.field not in kwargs.keys():
				raise Exception('%s is a required field' % self.field)
			if kwargs[self.field] not in self.test:
				raise Exception('%s is not a valid value (%s)' % (kwargs[self.field], str(self.test)))
			return func(*args, **kwargs)
		return wrapped

class PassiveTotal(object):
	"""Python helper library built on top of the PassiveTotal (www.passivetotal.org)
	service API. For updated documentation, head over to www.passivetotal.org/api.
	"""	
	def __init__(self, api_key):
		"""Initialize our object with a couple helpers and known-states
		@param	api_key	string	valid API key for PassiveTotal
		"""
		self._api_key = api_key
		
		self._endpoint = 'https://www.passivetotal.org/api/v1/'
		self._valid_methods = ['GET', 'POST']
		self._generic_get_types = ['metadata', 'passive', 'subdomains', \
							 'unique', 'classification', 'sinkhole' \
							 'dynamic', 'ever_compromised', 'watching']
		self._generic_post_types = ['classification', 'sinkhole', 'dynamic' \
									'ever_compromised', 'watching']
		self._classifications = ['targeted', 'crime', 'multiple', 'benign']
		self._logger = logging.getLogger('PassiveTotal')
		
	@property
	def endpoint(self):
		"""Endpoint getter"""
		return self._endpoint
		
	@endpoint.setter
	def endpoint(self, endpoint):
		"""PassiveTotal offers different versions of the API, so a user
		could in theory could be using something beyond the latest. 
		
		@param	endpoint	string	fully qualified URL of the API
		@returns	bool	whether or not the setting was successful
		"""
		if endpoint.split(':')[0].lower() not in ['http', 'https']:
			raise Exception('Endpoint must start with HTTP/HTTPS')
		if not endpoint.endswith('/'):
			endpoint += '/'
		self._endpoint = endpoint
		
		return True
	
	@property
	def logger(self):
		"""Logger getter"""
		return self._logger.getEffectiveLevel()
		
	@logger.setter	
	def logger(self, level):
		"""Simple logger to provide feedback when using the library"""
		logger = logging.getLogger('PassiveTotal')
		if level == "INFO":
			logger.setLevel(logging.INFO)
		elif level == "WARN":
			logger.setLevel(logging.WARN)
		elif level == "DEBUG":
			logger.setLevel(logging.DEBUG)
		elif level == "ERROR":
			logger.setLevel(logging.ERROR)
		else:
			raise Exception('%s was not a valid logging level' % level)
		format = logging.Formatter('\033[1;32m%(levelname)-5s %(module)s:%(funcName)s():%(lineno)d %(asctime)s\033[0m| %(message)s')
		shandler = logging.StreamHandler(sys.stdout)
		shandler.setFormatter(format)
		logger.addHandler(shandler)
		return logger
		
	# Pass-through GET methods
	def get_metadata(self, query_value):
		return self._router('GET', 'metadata', query_value)
	def get_passive(self, query_value):
		return self._router('GET', 'passive', query_value)
	def get_subdomains(self, query_value): 
		return self._router('GET', 'subdomains', query_value)
	def get_unique(self, query_value):
		return self._router('GET', 'unique', query_value)
	def get_classification(self, query_value):
		return self._router('GET', 'classification', query_value)
	def get_user_tags(self, query_value):
		return self._router('GET', 'user/tags', query_value)
	def get_sinkhole(self, query_value):
		return self._router('GET', 'sinkhole', query_value)
	def get_ever_compromised(self, query_value):
		return self._router('GET', 'ever_compromised', query_value)
	def get_dynamic(self, query_value):
		return self._router('GET', 'dynamic', query_value)
	def get_watching(self, query_value):
		return self._router('GET', 'watching', query_value)
		
	# Pass-through POST methods
	@set_check('classification', ['targeted', 'crime', 'multiple', 'benign'])
	def set_classification(self, query_value, **kwargs):
		return self._router('POST', 'classification', query_value, kwargs)
	@set_check('sinkhole', ['true', 'false'])
	def set_sinkhole(self, query_value, **kwargs):
		return self._router('POST', 'sinkhole', query_value, kwargs)
	@set_check('dynamic', ['true', 'false'])
	def set_dynamic(self, query_value, **kwargs):
		return self._router('POST', 'dynamic', query_value, kwargs)	
	@set_check('ever_compromised', ['true', 'false'])
	def set_ever_compromised(self, query_value, **kwargs):
		return self._router('POST', 'ever_compromised', query_value, kwargs)
	@set_check('watching', ['true', 'false'])
	def set_watching(self, query_value, **kwargs):
		return self._router('POST', 'watching', query_value, kwargs)
	def add_tag(self, query_value, **kwargs):
		if 'tag' not in kwargs:
			raise Exception('Tag is a required field')
		return self._router('POST', 'user/tag/add', query_value, kwargs)
	def remove_tag(self, query_value, **kwargs):
		if 'tag' not in kwargs:
			raise Exception('Tag is a required field')
		return self._router('POST', 'user/tag/remove', query_value, kwargs)
	
	# Helper methods	
	def _router(self, method, query_type, query_value, kwargs=None):
		"""Generic routing method to get the data to and from
		PassiveTotal. All methods should route through this.
		
		@param	method		string	GET or POST
		@param	query_type	string	a valid query endpoint
		@param	query_value	string	item to get data for
		@returns	dict	loaded JSON response
		"""
		call_url = self._endpoint + query_type
		self._logger.debug('Calling: %s' % call_url)
		params = {'api_key': self._api_key, 'query': query_value}
		self._logger.debug('Params: %s' % str(params))
		if method == 'GET':
			response = requests.get(call_url, params=params, verify=False)
		else:
			params.update(kwargs) # update our dict
			response = requests.post(call_url, params=params, verify=False)
		self._logger.debug('Response: %s' % str(response))
		json_response = json.loads(response.content)
		self._logger.debug('Loaded JSON: %s' % json_response)
		return json_response

File: api_helpers/php/passivetotal.php
<?php

	/**
	 * PHP helper library built on top of the PassiveTotal (www.passivetotal.org)
	 * service API. For updated documentation, head over to www.passivetotal.org/api.
	*/ 
	class PassiveTotal {
	
		/**
		 * Initialize our object with a couple helpers and known-states
		 *
		 * @param	$api_key	string	API key from the PassiveTotal website
		*/
		public function __construct($api_key) {
		    $this->api_key = $api_key;
		    
		    $this->endpoint = 'https://www.passivetotal.org/api/v1/';
			$this->valid_classifications = array('targeted', 'crime', 'multiple', 'benign');
		}
		
		/** 
		 * Generic getter methods that use the router pass-through
		*/
		public function getMetadata($query) {
			return $this->router('GET', 'metadata', $query);
		}
		public function getPassive($query) {
			return $this->router('GET', 'passive', $query);
		}
		public function getSubdomains($query) {
			return $this->router('GET', 'subdomains', $query);
		}
		public function getUnique($query) {
			return $this->router('GET', 'unique', $query);
		}
		public function getClassification($query) {
			return $this->router('GET', 'classification', $query);
		}
		public function getUserTags($query) {
			return $this->router('GET', 'user/tags', $query);
		}
		public function getSinkhole($query) {
			return $this->router('GET', 'sinkhole', $query);
		}
		public function getEverCompromised($query) {
			return $this->router('GET', 'ever_compromised', $query);
		}
		public function getDynamic($query) {
			return $this->router('GET', 'dynamic', $query);
		}
		public function getWatching($query) {
			return $this->router('GET', 'watching', $query);
		}
		
		/** 
		 * Generic setter methods that use the router pass-through
		*/
		public function setClassification($query, array $kwargs) {
			if (!array_key_exists('classification', $kwargs)) {
				throw new Exception('Classification field is required');
			}
			if (!in_array($kwargs['classification'], $this->valid_classifications)) {
				throw new exception('Classification type is not valid');
			}
			return $this->router('POST', 'classification', $query, $kwargs);
		}
		public function setSinkhole($query, array $kwargs) {
			if (!array_key_exists('sinkhole', $kwargs)) {
				throw new Exception('Sinkhole field is required');
			}
			return $this->router('POST', 'sinkhole', $query, $kwargs);
		}
		public function setDynamic($query, array $kwargs) {
			if (!array_key_exists('dynamic', $kwargs)) {
				throw new Exception('Dynamic field is required');
			}
			return $this->router('POST', 'dynamic', $query, $kwargs);
		}
		public function setEverCompromised($query, array $kwargs) {
			if (!array_key_exists('ever_compromised', $kwargs)) {
				throw new Exception('Ever_compromised field is required');
			}
			return $this->router('POST', 'ever_compromised', $query, $kwargs);
		}
		public function setWatching($query, array $kwargs) {
			if (!array_key_exists('watching', $kwargs)) {
				throw new Exception('Watching field is required');
			}
			return $this->router('POST', 'watching', $query, $kwargs);
		}
		public function addUserTag($query, array $kwargs) {
			if (!array_key_exists('tag', $kwargs)) {
				throw new Exception('Tag field is required');
			}
			return $this->router('POST', 'user/tag/add', $query, $kwargs);
		}
		public function removeUserTag($query, array $kwargs) {
			if (!array_key_exists('tag', $kwargs)) {
				throw new Exception('Tag field is required');
			}
			return $this->router('POST', 'user/tag/remove', $query, $kwargs);
		}
		
		/**
		 * Helper function to route user requests to the proper API endpoint
		 * handler. Results will always return JSON.
	     *
	     * @param	$method			string	GET or POST
	     * @param	$query_type		string	Endpoint to query
	     * @param	$query_value	string	Item to query
	     * @param	$kwargs			array	POST variables if needed
	     * @returns JSON encoded data back to the client
		*/
		private function router($method, $query_type, $query_value, array $kwargs=null) {
			$ch = curl_init(); 
		
			$params = array(
				'api_key' => $this->api_key,
				'query' => $query_value
			);
			
			if ($method === 'GET') {
				$query_string = http_build_query($params);
			} else {
				$full_url = $this->endpoint . $query_type;
				$post_fields = array_merge($params, $kwargs);
				$query_string = http_build_query($post_fields);
				curl_setopt($ch,CURLOPT_POST, count($post_fields));
				curl_setopt($ch,CURLOPT_POSTFIELDS, $query_string);	
			}
			
			$full_url = $this->endpoint . $query_type . '?' . $query_string;
			
		    curl_setopt($ch, CURLOPT_URL, $full_url); 
		    curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 0);
		    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, 0);
		    
		    $output = curl_exec($ch); 
		    curl_close($ch);  
		    
		    return json_decode($output);
		}
	}
?>
